org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/utils/NodeJsManager.java

```diff
@@ -0,0 +1,94 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarlint.eclipse.core.internal.utils;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Objects;
+import org.eclipse.jdt.annotation.Nullable;
+import org.sonarlint.eclipse.core.internal.SonarLintCorePlugin;
+import org.sonarlint.eclipse.core.internal.engine.AnalysisRequirementNotifications;
+import org.sonarlint.eclipse.core.internal.engine.connected.ConnectedEngineFacade;
+import org.sonarlint.eclipse.core.internal.preferences.SonarLintGlobalConfiguration;
+import org.sonarsource.sonarlint.core.NodeJsHelper;
+import org.sonarsource.sonarlint.core.client.api.common.Version;
+
+public class NodeJsManager {
+
+  private boolean nodeInit = false;
+  private Path nodeJsPath = null;
+  private Version nodeJsVersion = null;
+
+  /**
+   * Reload path from global preferences.
+   * Should be called when preferences are changed.
+   */
+  public void reload() {
+    if (!Objects.equals(Paths.get(SonarLintGlobalConfiguration.getNodejsPath()), nodeJsPath)) {
+      clear();
+      // Node.js path is passed at engine startup, so we have to restart them all to ensure the new value is taken into account
+      SonarLintCorePlugin.getInstance().getDefaultSonarLintClientFacade().stop();
+      SonarLintCorePlugin.getServersManager().getServers().forEach(f -> ((ConnectedEngineFacade) f).stop());
+      AnalysisRequirementNotifications.resetCachedMessages();
+    }
+  }
+
+  private synchronized void clear() {
+    this.nodeInit = false;
+    this.nodeJsPath = null;
+    this.nodeJsVersion = null;
+  }
+
+  private synchronized void initNodeIfNeeded() {
+    if (!nodeInit) {
+      NodeJsHelper helper = new NodeJsHelper();
+      helper.detect(getNodeJsPathFromConfig());
+      this.nodeInit = true;
+      this.nodeJsPath = helper.getNodeJsPath();
+      this.nodeJsVersion = helper.getNodeJsVersion();
+    }
+  }
+
+  @Nullable
+  public Path getNodeJsPath() {
+    initNodeIfNeeded();
+    return nodeJsPath;
+  }
+
+  @Nullable
+  public Version getNodeJsVersion() {
+    initNodeIfNeeded();
+    return nodeJsVersion;
+  }
+
+  @Nullable
+  private static Path getNodeJsPathFromConfig() {
+    final String nodejsPathStr = SonarLintGlobalConfiguration.getNodejsPath();
+    if (StringUtils.isNotBlank(nodejsPathStr)) {
+      try {
+        return Paths.get(nodejsPathStr);
+      } catch (Exception e) {
+        throw new IllegalStateException("Invalid Node.js path", e);
```

Comment:
I am trying to see if this exception would be caught at some point up in the call hierarchy and I'm not sure it would be the case for every caller. I think this exception happens if the path is malformed, so it is maybe already checked in the settings popup (in `checkState`) ?

Reply:
Do you mean I should remove the catch block?